### PR TITLE
[SPARK-39368][SQL] Move `RewritePredicateSubquery` into `InjectRuntimeFilter`

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR moves `RewritePredicateSubquery` into `InjectRuntimeFilter`.

### Why are the changes needed?

Reduce the number of `RewritePredicateSubquery` runs, since `spark.sql.optimizer.runtimeFilter.semiJoinReduction.enabled` is disabled by default. For example:
```
build/sbt "sql/testOnly *TPCDSQuerySuite"
```
Before this PR:
```
...
org.apache.spark.sql.catalyst.optimizer.RewritePredicateSubquery                      17978319 / 31026106                             26 / 624
...
```
After this PR:
```
...
org.apache.spark.sql.catalyst.optimizer.RewritePredicateSubquery                      16680901 / 18994542                             26 / 312
...
```

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing test.

### CODE REVIEW

**Comment by wangyum** [comment]
cc @sigmod @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InjectRuntimeFilter.scala [hunk 1]
```diff
@@ -288,7 +288,13 @@ object InjectRuntimeFilter extends Rule[LogicalPlan] with PredicateHelper with J
     case s: Subquery if s.correlated => plan
     case _ if !conf.runtimeFilterSemiJoinReductionEnabled &&
       !conf.runtimeFilterBloomFilterEnabled => plan
-    case _ => tryInjectRuntimeFilter(plan)
+    case _ =>
+      val newPlan = tryInjectRuntimeFilter(plan)
+      if (conf.runtimeFilterSemiJoinReductionEnabled) {
```
Comment (sigmod) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InjectRuntimeFilter.scala:293]: how about 

`if (conf.runtimeFilterSemiJoinReductionEnable && !plan.fastEquals(newPlan))`

**Review by sigmod — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InjectRuntimeFilter.scala (hunk 1 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InjectRuntimeFilter.scala:293]: +1

**Review by sigmod — APPROVED** [review]

**Review by singhpk234 — APPROVED** [review]

**Comment by wangyum** [comment]
Merged to master.

**Pull request closed without merging**